features/lorawan/LoRaWANStack.cpp

```diff
@@ -0,0 +1,2046 @@
+        }
+
+        // skip any channels which are not currently enabled
+        if ((channel_masks[0] & (1U<<i)) == 0) {
+            continue;
+        }
+
+        status = error_type_converter(LoRaMacChannelRemove(i));
+
+        if (status != LORA_MAC_STATUS_OK) {
+            return status;
+        }
+    }
+
+    return status;
+}
+
+lora_mac_status_t LoRaWANStack::remove_a_channel(uint8_t channel_id)
+{
+    if (_device_current_state == DEVICE_STATE_NOT_INITIALIZED )
+    {
+        tr_error("Stack not initialized!");
+        return LORA_MAC_STATUS_NOT_INITIALIZED;
+    }
+
+    GetPhyParams_t get_phy;
+    PhyParam_t phy_param;
+    uint8_t max_num_channels;
+    uint16_t *channel_masks;
+
+    // Check first how many channels the selected PHY layer supports
+    get_phy.Attribute = PHY_MAX_NB_CHANNELS;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    max_num_channels = (uint8_t) phy_param.Value;
+
+    // According to specification channel IDs start from 0 and last valid
+    // channel ID is N-1 where N=MAX_NUM_CHANNELS.
+    // So any ID which is larger or equal to the Max number of channels is invalid
+    if (channel_id >= max_num_channels) {
+        return LORA_MAC_STATUS_PARAMETER_INVALID;
+    }
+
+    // Now check the Default channel mask
+    get_phy.Attribute = PHY_CHANNELS_DEFAULT_MASK;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    channel_masks = phy_param.ChannelsMask;
+
+    // check if the channel ID give belongs to a default channel
+    // Mostly the default channels are in the first mask if the region
+    // have multiple channel masks for various sub-bands. So we check the first
+    // mask only and return an error code if user sent a default channel id
+    if ((channel_masks[0] & (1U << channel_id)) != 0) {
+        tr_error("Not allowed to remove a Default Channel.");
+        return LORA_MAC_STATUS_PARAMETER_INVALID;
+    }
+
+    return error_type_converter(LoRaMacChannelRemove(channel_id));
+}
+
+lora_mac_status_t LoRaWANStack::get_enabled_channels(lora_channelplan_t& channel_plan)
+{
+    if (_device_current_state == DEVICE_STATE_JOINING
+            || _device_current_state == DEVICE_STATE_NOT_INITIALIZED
+            || _device_current_state == DEVICE_STATE_INIT)
+    {
+        tr_error("Cannot get channel plan until Joined!");
+        return LORA_MAC_STATUS_BUSY;
+    }
+
+    lora_mac_mib_request_confirm_t mib_params;
+
+    GetPhyParams_t get_phy;
+    PhyParam_t phy_param;
+    uint8_t max_num_channels;
+    uint16_t *channel_masks;
+    uint8_t count = 0;
+
+    // Check first how many channels the selected PHY layer supports
+    get_phy.Attribute = PHY_MAX_NB_CHANNELS;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    max_num_channels = (uint8_t) phy_param.Value;
+
+    // Now check the Default channel mask
+    get_phy.Attribute = PHY_CHANNELS_MASK;
+    phy_param = lora_phy.get_phy_params(&get_phy);
+    channel_masks = phy_param.ChannelsMask;
```

Comment:
Better to check if channel_mask is NULL?

Reply:
Same comment as above